### ISSUE TITLE
Duplicated Many/Many Relationships

### ISSUE THREAD

**Issue by smallhadroncollider** [analogueorm/analogue#139]
I feel like I've reported something similar before, but I couldn't find it...

---

I have `Score`s which can have many `ProductionCompany`s and vice versa:

```php
class ScoreMap extends EntityMap
{
    public function production_companies(Score $score)
    {
        return $this->belongsToMany($score, ProductionCompany::class, "production_companies_scores", "score_id", "production_company_id");
    }
}

class ProductionCompanyMap extends EntityMap
{
    public function scores(ProductionCompany $productionCompany)
    {
        return $this->belongsToMany($productionCompany, Score::class, "production_companies_scores", "production_company_id", "score_id");
    }
}
```

When I store a `Score`, it creates multiple copies of the relationship in the database. Every time a `Score` gets updated the relationship seems to duplicate itself. This can get quite out of hand quite quickly.

**Comment by RemiCollin** [comment]
Yes I think I remember something like this. Can you paste me the code where the Score get updated and persisted ?

**Comment by smallhadroncollider** [comment]
Actually, looking at it, the setup is a little bit more complex.

The `Scores` are never stored directly, always via another type of `Entity`. E.g. a `Film` is a `Score`, so when a `Film` is first created a new `Score` is also created an then linked to. `Production Companies` belong to a `Score`, so they get added to the `Score`, but are saved when the `Film` is stored (so, `Score`s are never stored directly).

```php
    // in the FilmFactory class
    public function create()
    {
        $entity = new $this->class;
        $entity->score = (new ScoreFactory)->create();
        $entity->score->scoreable = $entity;
        return $entity;
    }
```

Then, storing the Film:

```php
$this->filmRepository->store($film);
```
Where `$this->filmRepository` is just a standard `Repository`.

The `Film`'s `Score` definitely has the correct (i.e. no duplicates) data just before it's stored.

Perhaps it's the fact it is saved via another Entity that's causing the issue, as I just checked and it doesn't seem to affect the other Entities with many-many relationships.

**Comment by RemiCollin** [comment]
Hard to say where it comes from at first glance. I'm wondering if the inverse of the relationshing might be the cause of the issue. I'll have to unit test some similar edge case to find the real cause.

**Comment by RemiCollin** [comment]
Closing this has it was related to a bug in the cache which was fixed in 5.5
